artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/protocol/amqp/proton/AMQPLargeMessageReader.java

```diff
@@ -77,39 +74,68 @@ public AMQPLargeMessageReader open() {
+      logger.trace("Reading {}", delivery);
+
       final Receiver receiver = ((Receiver) delivery.getLink());
       final ReadableBuffer dataBuffer = receiver.recv();
 
+      final AMQPSessionCallback sessionSPI = serverReceiver.getSessionContext().getSessionSPI();
+
       if (currentMessage == null) {
-         final AMQPSessionCallback sessionSPI = serverReceiver.getSessionContext().getSessionSPI();
          final long id = sessionSPI.getStorageManager().generateID();
-         currentMessage = new AMQPLargeMessage(id, delivery.getMessageFormat(), null,
-                                               sessionSPI.getCoreMessageObjectPools(),
-                                               sessionSPI.getStorageManager());
+         currentMessage = new AMQPLargeMessage(id, delivery.getMessageFormat(), null, sessionSPI.getCoreMessageObjectPools(), sessionSPI.getStorageManager());
          currentMessage.parseHeader(dataBuffer);
-
+         logger.trace("Initializing current message {} on {}", currentMessage, this);
          sessionSPI.getStorageManager().largeMessageCreated(id, currentMessage);
+         sessionSPI.execute(currentMessage::validateFile);
       }
 
-      currentMessage.addBytes(dataBuffer);
+      logger.trace("Disable autoread on {}", this);
+      serverReceiver.getConnection().disableAutoRead();
 
-      final AMQPLargeMessage result;
+      byte[] bytes = new byte[dataBuffer.remaining()];
+      dataBuffer.get(bytes);
+      logger.trace("Getting {}", bytes.length);
 
-      if (!delivery.isPartial()) {
-         currentMessage.releaseResources(serverReceiver.getConnection().isLargeMessageSync(), true);
-         result = currentMessage;
-         // We don't want a close to delete the file now, we've released the resources.
-         currentMessage = null;
-         deliveryAnnotations = result.getDeliveryAnnotations();
-      } else {
-         result = null;
-      }
+      boolean partial = delivery.isPartial();
+
+      sessionSPI.execute(() -> addBytes(delivery, bytes, partial));
+
+      return partial;
+   }
 
-      return result;
+   private void addBytes(Delivery delivery, byte[] bytes, boolean isPartial) {
+      ReadableBuffer dataBuffer = ReadableBuffer.ByteBufferReader.wrap(bytes);
+      try {
+         logger.trace("Adding {} bytes on currentMessage={}, this={}", dataBuffer.remaining(), currentMessage, this);
+         currentMessage.addBytes(dataBuffer);
+
+         if (!isPartial) {
+            final AMQPLargeMessage message = currentMessage;
+            message.releaseResources(serverReceiver.getConnection().isLargeMessageSync(), true);
+            logger.trace("finishing {} on {}", currentMessage, this);
+            // We don't want a close to delete the file now, we've released the resources.
+            currentMessage = null;
+            close();
+            serverReceiver.connection.runNow(() -> {
+               DeliveryAnnotations deliveryAnnotations = message.getDeliveryAnnotations();
```

Comment:
Can skip temp var creation and just call message.getDeliveryAnnotations directly.